File: ProjectActivityLD2/app/src/main/java/com/example/projectactivityld/CustomAdapterLocation.java
package com.example.projectactivityld;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import java.util.List;

public class CustomAdapterLocation extends ArrayAdapter<Location> {

    public CustomAdapterLocation(Context context, int textViewResourceId,
                                 List<Location> objects) {
        super(context, textViewResourceId, objects);

    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        return getViewOptimize(position, convertView, parent);
    }

    public View getViewOptimize(int position, View convertView, ViewGroup parent) {
        ViewHolder viewHolder = null;
        if (convertView == null) {
            LayoutInflater inflater = (LayoutInflater) getContext()
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            convertView = inflater.inflate(R.layout.row_layout, null);
            viewHolder = new ViewHolder();
            viewHolder.name = (TextView)convertView.findViewById(R.id.textViewName);

            convertView.setTag(viewHolder);

        } else {
            viewHolder = (ViewHolder) convertView.getTag();
        }

        Location location = getItem(position);

        viewHolder.name.setText(location.getName());

        return convertView;
    }



    private class ViewHolder {

        public TextView name;

    }

}

File: ProjectActivityKot/app/src/main/java/com/example/projectactivitykot/CustomAdapterLocation.kt
package com.example.projectactivitykot

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView

class CustomAdapterLocation(context: Context, textViewResourceId: Int, objects: List<Location>) : ArrayAdapter<Location>(context, textViewResourceId, objects) {

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        return getViewOptimize(position, convertView, parent)
    }

    fun getViewOptimize(position: Int, convertView: View?, parent: ViewGroup): View {

        var convertView = convertView
        var viewHolder: ViewHolder? = null
        if (convertView == null) {
            val inflater = context
                .getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            convertView = inflater.inflate(R.layout.row_layout, null)
            viewHolder = ViewHolder()
            viewHolder.name = convertView!!.findViewById<View>(R.id.textViewName) as TextView

            convertView.tag = viewHolder

        } else {
            viewHolder = convertView.tag as ViewHolder
        }
        val location = getItem(position)
        viewHolder.name!!.setText(location!!.name)

        return convertView
    }


    private inner class ViewHolder {

        var name: TextView? = null
    }

}